notebooks/source/bayesian_hierarchical_linear_regression.ipynb

```diff
@@ -0,0 +1,677 @@
+    "\n",
+    "Mathematically, the model is described by the following equations:\n",
+    "$$\n",
+    "\\mu_{\\alpha} \\sim \\mathcal{N}(0, 100) \\\\\n",
+    "\\sigma_{\\alpha} \\sim |\\mathcal{N}(0, 100)| \\\\\n",
+    "\\mu_{\\beta} \\sim \\mathcal{N}(0, 100) \\\\\n",
+    "\\sigma_{\\beta} \\sim |\\mathcal{N}(0, 100)| \\\\\n",
+    "\\alpha_i \\sim \\mathcal{N}(\\mu_{\\alpha}, \\sigma_{\\alpha}) \\\\\n",
+    "\\beta_i \\sim \\mathcal{N}(\\mu_{\\beta}, \\sigma_{\\beta}) \\\\\n",
+    "\\sigma \\sim \\mathcal{N}(0, 100) \\\\\n",
+    "FVC_{ij} \\sim \\mathcal{N}(\\alpha_i + t \\beta_i, \\sigma)\n",
+    "$$\n",
+    "\n",
+    "where *t* is the time in weeks. Those are very uninformative priors, but that's ok: our model will converge!\n",
+    "\n",
+    "Implementing this model in Numpyro is pretty straightforward:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpyro\n",
+    "from numpyro.infer import MCMC, NUTS, Predictive\n",
+    "import numpyro.distributions as dist\n",
+    "from jax import random"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def model(PatientID, Weeks, FVC_obs=None):\n",
+    "    mu_a = numpyro.sample(\"mu_a\", dist.Normal(0., 100.))\n",
+    "    sigma_a = numpyro.sample(\"sigma_a\", dist.HalfNormal(100.))\n",
+    "    mu_b = numpyro.sample(\"mu_b\", dist.Normal(0., 100.))\n",
+    "    sigma_b = numpyro.sample(\"sigma_b\", dist.HalfNormal(100.))\n",
+    "    \n",
+    "    unique_patient_IDs = np.unique(PatientID)\n",
+    "    n_patients = len(unique_patient_IDs)\n",
+    "    \n",
+    "    a = numpyro.sample(\"a\", dist.Normal(mu_a, sigma_a).expand([n_patients]))\n",
+    "    b = numpyro.sample(\"b\", dist.Normal(mu_b, sigma_b).expand([n_patients]))\n",
+    "    \n",
+    "    sigma = numpyro.sample(\"sigma\", dist.HalfNormal(100.))\n",
+    "    \n",
+    "    FVC_est = a[PatientID] + b[PatientID] * Weeks\n",
+    "    \n",
+    "    with numpyro.plate(\"data\", len(PatientID)):\n",
+    "        numpyro.sample(\"obs\", dist.Normal(FVC_est, sigma), obs=FVC_obs)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "That's all for modelling!\n",
+    "\n",
+    "## 3. Fitting the model\n",
+    "A great achievement of Probabilistic Programming Languages such as Numpyro is to decouple model specification and inference. After specifying my generative model, with priors, condition statements and data likelihood, I can leave the hard work to Numpyro's inference engine. \n",
+    "\n",
+    "Calling it requires just a few lines. Before we do it, let's add a numerical Patient ID:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.preprocessing import LabelEncoder\n",
+    "\n",
+    "le = LabelEncoder()\n",
```

Comment:
Could you add a comment on why we need to do this preprocessing step? It is trivial for sklearn folks but might be not with other readers.